Improvements to allocation performance on machines with many processors

PeterSolMS opened the pull request:
In particular:

- Make sure we reserve memory on the correct NUMA node. Otherwise the OS will pick a NUMA node for us on first touch, which is sometimes correct, sometimes not.

- Only look at a subset of the available heaps in balance_heaps. Look at a different subset each time, and look both at heaps inside and outside the allocating processor's NUMA node, but give preference to heaps within the allocating processor's NUMA node and in particular, the heap associated with the allocating processor.
- Fix issue where our logic assumed that NUMA node numbers are non-decreasing as the processor numbers increase, but this doesn't hold true on all machines.

PeterSolMS committed:
Improvements to allocation performance on machines with many processors. In particular:
Make sure we reserve memory on the correct NUMA node. Otherwise the OS will pick a NUMA node for us on first touch, which is sometimes correct, sometimes not.
Only look at a subset of the available heaps in balance_heaps. Look at a different subset each time, and look both at heaps inside and outside the allocating processor's NUMA node, but give preference to heaps within the allocating processor's NUMA node and in particular, the heap associated with the allocating processor.
Fix issue where our logic assumed that NUMA node numbers are non-decreasing as the processor numbers increase, but this doesn't hold true on all machines.

Maoni0 commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -14237,84 +14414,153 @@ void gc_heap::balance_heaps (alloc_context* acontext)
-                        }
-                        int hp_alloc_context_count = hp->alloc_context_count;
+                        // favor new home heap over org heap
+                        size += delta * 2;
```
so we add delta*2 instead of delta for new home heap. what was the reason for this? 

also, what if the new home heap is on a proc on a remote node?

PeterSolMS replied:
This gave slightly better performance.

You mean "remote" in the sense that it's on a different node than the org heap? I think in that case it's likely better to switch, because as the new home heap is on the same node as the processor we're executing on, we should see better memory bandwidth.

Maoni0 commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -14237,84 +14414,153 @@ void gc_heap::balance_heaps (alloc_context* acontext)
+                        // alloc_count often increases by multiples of 16 (due to logic at top of routine),
+                        // and we want to advance the starting point by 4 between successive calls,
+                        // therefore the shift right by 2 bits
+                        int heap_num = start + ((acontext->alloc_count >> 2) + new_home_hp_num) % count;
```
we could be looking at new home heap again here and would undo what we just did above.

PeterSolMS replied:
I don't get the "undo" part - it's true we could look at new home heap again, but why would it hurt (other than being a waste of time)?

Maoni0 replied:
ahh you are right, it won't undo 'cause if the first check doesn't change max_size, the 2nd check (with the size not being delta*2 larger) certain wouldn't. 

I'm wondering why you made the code this way instead of the original way which seems more readable to me...

PeterSolMS replied:
The problem was that as I am only testing a subset of the heaps, there was no guarantee that the new home heap would even be considered. At first I tried to shoehorn it into the loop, but that seemed awkward.

Maoni0 commented on src/coreclr/src/gc/env/gcenv.os.h:
```diff
@@ -290,7 +290,7 @@ class GCToOSInterface
     // Return:
     //  Address of the allocated memory
-    static void* VirtualReserveAndCommitLargePages(size_t size);
+    static void* VirtualReserveAndCommitLargePages(size_t size, uint16_t node = NUMA_NODE_UNDEFINED);
```
since you are changing this you'll need to change the linux impl as well in src\gc\unix\gcenv.unix.cpp

PeterSolMS replied:
Yes, good point. Will do.

Maoni0 commented:
also in `virtual_alloc_commit_for_heap` we shouldn't bother with the numa node unless it's still useful for Linux.

PeterSolMS commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -4428,88 +4440,219 @@ BOOL reserve_initial_memory (size_t normal_size, size_t large_size, int num_heap
 
-        if (b1 && b2)
+            // add the size required for this heap
+            block->block_size += normal_size + large_size;
```
I see, I need to catch up with POH - will do.

PeterSolMS replied:
Ok, I took care of that.

PeterSolMS commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -4383,11 +4392,14 @@ typedef struct
 initial_memory_details memory_details;
 
-BOOL reserve_initial_memory (size_t normal_size, size_t large_size, int num_heaps, bool use_large_pages_p)
+BOOL reserve_initial_memory (size_t normal_size, size_t large_size, int num_heaps, bool use_large_pages_p, uint16_t heap_no_to_numa_node[])
```
Ok. My reason was to indicate this is actually an array...

PeterSolMS replied:
Done.

PeterSolMS commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -5312,10 +5471,10 @@ class heap_select
-        // numa_node_to_heap_map[numa_node] is set to the first heap number on that node and
-        // numa_node_to_heap_map[numa_node + 1] is set to the first heap number not on that node
+        // numa_node_to_heap_start[numa_node] is set to the first heap number on that node and
+        // numa_node_to_heap_last[numa_node] is set to the first heap number not on that node
```
Ok.

PeterSolMS replied:
Thanks for spotting this!

PeterSolMS committed:
Merge with Vlad's changes.

PeterSolMS committed:
Address some code review feedback, changes for pinned heap.

PeterSolMS commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -4428,88 +4440,219 @@ BOOL reserve_initial_memory (size_t normal_size, size_t large_size, int num_heap
+        assert (highest_numa_node < MAX_SUPPORTED_CPUS);
+
+        memory_details.numa_reserved_block_count = highest_numa_node + 1;
+        memory_details.numa_reserved_block_table = new (nothrow) numa_reserved_block[memory_details.numa_reserved_block_count];
```
I have a default constructor on numa_reserved_block that should take care of zeroing out the memory.

PeterSolMS commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -4593,7 +4751,7 @@ void* virtual_alloc (size_t size, bool use_large_pages_p)
 
     void* prgmem = use_large_pages_p ?
         GCToOSInterface::VirtualReserveAndCommitLargePages(requested_size) :
```
No, that's just an oversight. Let me put in the parameter and do a test to make sure it works.

PeterSolMS replied:
Ok, there was another place where the numa node parameter wasn't getting passed on, I took care of that as well.

PeterSolMS committed:
Address numa node parameter for large pages at the call site and the linux implementation.

Maoni0 commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -4550,24 +4695,23 @@ void destroy_initial_memory()
+            virtual_free (memory_details.initial_memory[0].memory_base,
+                memory_details.block_count * (memory_details.block_size_normal +
+                    memory_details.block_size_large));
         }
```
you'd also want to add the pinned seg size here if you want to be consistent...even though this does not actually get called. 

nit - it seems a little odd that you made some formatting changes but only for part of this method...

PeterSolMS replied:
Ah, this was existing code that was missing pinned seg size. I put it in.

Some of the formatting changes were by mistake, only the space after virtual_free was intentional.

PeterSolMS committed:
Undo accidental formatting changes, account for block_size_inned in destroy_initial_memory, implement large pages with NUMA support.

PeterSolMS committed:
Fix assert failure on x86 - need to pass write watch flag to VirtualAllocExNuma.

PeterSolMS committed:
Merge branch 'master' into heap_balance_improvements

PeterSolMS committed:
Merge with master

PeterSolMS committed:
Enforce contiguous range of heap numbers for each NUMA node, undo earlier incomplete fix for this issue.

PeterSolMS committed:
Merge with master, will need to straighten out reserve_initial_memory.

PeterSolMS committed:
Update gc_heap::reserve_initial_memory to honor separated_poh_p in the NUMA case.

PeterSolMS committed:
Fixed for-loop bug in reserve_initial_memory, changed if-else if in destroy_initial_memory to a switch to make it more readable, fixed interaction issue between POH hard limit and number of heaps.

cshung commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -4019,171 +4030,349 @@ BOOL gc_heap::reserve_initial_memory (size_t normal_size, size_t large_size, siz
-                memory_details.initial_pinned_heap[i].memory_base = separated_poh_block +
-                                                                  (i * pinned_size);
+                dprintf(2, ("failed to reserve %Id bytes for on NUMA node %u", block->block_size, numa_node));
+                failure = true;
```
If failure happens, `memory_details.numa_reserved_block_table[b].memory_base` will be garbage for `b` > `block_index`.

PeterSolMS replied:
Ah, yes, that's why I provided a default constructor for numa_reserved_block.:

struct numa_reserved_block
{
    uint8_t*        memory_base;
    size_t          block_size;

    numa_reserved_block() : memory_base(nullptr), block_size(0) { }
};

That way, the "new (nothrow) numa_reserved_block[memory_details.numa_reserved_block_count]" statement will call the default constructor on each element, and everything will be neatly 0-initialized.

cshung commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -4019,171 +4030,349 @@ BOOL gc_heap::reserve_initial_memory (size_t normal_size, size_t large_size, siz
+
+                if (block->memory_base != nullptr)
+                {
+                    virtual_free(block->memory_base, block->block_size);
```
... and then we will be freeing garbage here.

PeterSolMS replied:
and so every block that we didn't do a virtual_alloc on will still have memory_base == nullptr .

PeterSolMS committed:
Make indices into heap_hard_limit_oh symbolic, i.e. use soh, loh and poh instead of 0, 1 and 2. This is just for readability- no functional change intended.

Fixed parallel issue to the one I found with hard limit percentage zero for POH - if you explicitly give hardlimit size as 0 for POH, we have the same problem and I fixed it in the same way by leaving heap_hard_limit[POH] at 0, later not taking it into account for heap count computation, and setting pin_seg_size to the min size.

PeterSolMS committed:
Merge branch 'master' into heap_balance_improvements

PeterSolMS committed:
Merge branch 'master' into heap_balance_improvements
